ext/drreg/drreg.c

```diff
@@ -2140,17 +2201,33 @@ drreg_event_restore_state_with_ilist(void *drcontext, bool restore_memory,
                 aflags_native = true;
                 forget_existing_spill_slot_use(spill_slot_to_reg, AFLAGS_ALIAS_REG);
                 aflags_spill_reg = DR_REG_NULL;
+                gpr_to_non_drreg_slot_offs[GPR_IDX(AFLAGS_ALIAS_REG)] = NO_OFFS;
             } else {
                 /* Tool wrote aflags. Not native anymore. */
                 aflags_native = false;
             }
         }
-
-        /* Update book-keeping for gpr/aflags spill slot and aflags spill reg. Mark
-         * as native the gprs/aflags that were restored. All spills/restores are meta
-         * instrs.
+        /* Update book-keeping for gpr/aflags spill slot or aflags spill reg, and
+         * mark as native the gprs/aflags that were restored by drreg. All spills/restores
+         * are meta instrs.
          */
         if (!instr_is_app(inst)) {
+            /* There are multiple types of instrs that may spill or restore the gpr/aflags
+             * native value:
+             * - drreg spill or restore: detected using is_our_spill_or_restore
+             * - non-drreg spill or restore to mcontext slot or TLS slot not reserved by
+             *   drreg: if above is false, detected using instr_is_reg_spill_or_restore_ex
+             * - non-drreg spill or restore to stack: e.g. this may happen in some cases
```

Comment:
Detecting offset for stack spills (future work; not in this PR) seems tricky. For x86, it may perhaps be easier (keep a relative offset which starts at zero, which is incremented when we see a `push` and decremented on `pop`, and use that as offset when we see a stack spill), but for AArch64 it's much more complex. We'd need to look for ldp/stp with not just the stack pointer, but any reg that's copied from sp (something like `mov x0 <- sp+10` means that stores to `x0` are also essentially spills to stack), and also more complex addressing modes.

As commented, I see stack spills happening in some cases for clean call instrumentation on AArch64.

Reply:
This doesn't seem like a blocker in fixing i#4963 though. Also, the current restore logic works fine for the clean call in drcachesim.

Reply:
Decided to tackle stack spills in a separate PR.